casper/src/main/resources/RevVault.rho

```diff
@@ -27,178 +30,347 @@ in {
-                @Either!("fromBoolean <-", *authKeyValidCh, "Invalid AuthKey", *authKeyValidEitherCh) |
-
-                @Either!("productR <-", *revAddressValidEither, *amountNonNegative, *parametersOkCh) |
-                @Either!("productR <-", *parametersOkCh, *authKeyValidEitherCh, *parametersAndAuthOkCh) |
-
-                @Either!("flatMap <-", *parametersAndAuthOkCh, *split, *eitherPurseCh) |
-                for (_, retCh <- split) {
-                  new amountPurseCh in {
-                    purse!("split", amount, *amountPurseCh) |
-                    @Either!("fromSingletonList <-", *amountPurseCh, "Insufficient funds", *retCh)
-                  }
-                } |
+            @AuthKey!("check", *authKey, (*_revVault, ownRevAddress), *authKeyValidCh) |
+            @Either!("fromBoolean <-", *authKeyValidCh, "Invalid AuthKey", *authKeyValidEitherCh) |
 
-                @Either!("flatMap <-", *eitherPurseCh, *doDeposit, *ret) |
-                for (@p, retCh <- doDeposit) {
-                  @{revAddress | bundle0{*_revVault}}!("_deposit", p, *retCh)
-                }
+            @Either!("productR <-", *revAddressValidEither, *amountNonNegative, *parametersOkCh) |
+            @Either!("productR <-", *parametersOkCh, *authKeyValidEitherCh, *parametersAndAuthOkCh) |
+
+            @Either!("flatMap <-", *parametersAndAuthOkCh, *split, *eitherPurseCh) |
+            for (_, retCh <- split) {
+              new amountPurseCh in {
+                purse!("split", amount, *amountPurseCh) |
+                @Either!("fromSingletonList <-", *amountPurseCh, "Insufficient funds", *retCh)
               }
             } |
 
-            contract _depositTemplate(purse, depositPurse, retCh) = {
-              new amountCh, depositSuccessCh in {
-                depositPurse!("getBalance", *amountCh) |
-                for (@amount <- amountCh) {
+            @Either!("flatMap <-", *eitherPurseCh, *doDeposit, *ret) |
+            for (@p, retCh <- doDeposit) {
+              @{revAddress | bundle0{*_revVault}}!("_deposit", p, *retCh)
+            }
+          }
+        } |
+
+        contract _depositTemplate(purse, depositPurse, retCh) = {
+          new amountCh, depositSuccessCh in {
+            depositPurse!("getBalance", *amountCh) |
+            for (@amount <- amountCh) {
 
-                  purse!("deposit", amount, *depositPurse, *depositSuccessCh) |
-                  @Either!("fromBoolean <-", *depositSuccessCh, "BUG FOUND: purse deposit failed", *retCh)
+              purse!("deposit", amount, *depositPurse, *depositSuccessCh) |
+              @Either!("fromBoolean <-", *depositSuccessCh, "BUG FOUND: purse deposit failed", *retCh)
+            }
+          }
+        } |
+
+        // MultiSig
+        // Returns channel on which to invoke methods
+        contract RevVault(@"createMultiSig", @publicKeys, @authKeys, @quorumSize, ret) = {
```

Comment:
the way I imagined it was for MultiSig to be a separate contract, only wrapping (delegating to) RevVault, without knowing its internal details / sharing any names. Is there a reason why this is not possible?

Reply:
This code is the culprit: `@ListOps!("map", publicKeys, *_fromPublicKey, *authPubKeysCh)`

It can be an entirely separate if you don't expect a deployer authkey from RevVault to work in a multisig vault.

Reply:
Yes, totally. There's no need for auth-key reuse, and I'd claim it's safer and more intuitive for each authKey to only have one purpose. And it reduces coupling between RevVault and Multisig.

Reply:
Done.